### PR TITLE
load doc comments from xml files when available

### LINKED ISSUE #626: F# completion item documentation is incorrect.
F# completion documentation shows the type name

```
Microsoft.FSharp.Control.FSharpAsync`1[FSharp.Compiler.SourceCodeServices.FSharpToolTipText`1[System.String]]
```

instead of the actual content.

image

Issue stems from [here](https://github.com/dotnet/interactive/blob/86d3f806832b53425bb1c95d2d88437cd467f05d/src/Microsoft.DotNet.Interactive.FSharp/FSharpKernel.fs#L74-L76).

### LINKED ISSUE #628: Hover text does not contain documentation (C#)
image

### LINKED ISSUE #727: Hover text does not contain documentation (F#)
C# and F# hover text doesn't contain the appropriate HTML doc comment.

### PR DESCRIPTION
**Pull request by brettfo**
This loads doc comments from `.xml` files that are next to `.dll` files and ensures that data is passed through to the events.

For all the currently supported language services (completion, hover, sighelp) I've added tests that cover retrieving doc comments from (source, assembly reference, NuGet package).

Filed issue #1071 to track some edge cases around NuGet and C# doc comments.
Filed issue #1072 to track loading doc comments from BCL/Framework types.

Fixes #626 (added test).
Fixes #628 (added test).
Fixes #727 (added test).

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs [hunk 1]
```diff
@@ -115,5 +116,34 @@ public async Task signature_help_can_return_doc_comments_from_source(Language la
                 .Should()
                 .Be(expectedMethodDocumentation);
         }
+
+        [Theory]
+        [InlineData(Language.CSharp, "new SampleLibrary.SampleLibrary($$")]
+        //[InlineData(Language.FSharp, "let s = SampleLibrary.SampleLibrary($$")] // F# doesn't support signature help yet
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs:122]: Please uncomment this and put a `Skip` on it so it shows up in test runs as skipped. It's a little easier to keep track of that way.

File: dotnet-interactive.sln [hunk 2]
```diff
@@ -74,7 +74,9 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "docs", "docs", "{4DD50BCA-1
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.DotNet.Interactive.Http", "src\Microsoft.DotNet.Interactive.Http\Microsoft.DotNet.Interactive.Http.csproj", "{DE942DC3-B41B-45C5-9E01-3C4F042D3916}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.DotNet.Interactive.AspNetCore", "src\Microsoft.DotNet.Interactive.AspNetCore\Microsoft.DotNet.Interactive.AspNetCore.csproj", "{5167FC61-2FBF-4E0D-877E-C2DD354FB955}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.DotNet.Interactive.AspNetCore", "src\Microsoft.DotNet.Interactive.AspNetCore\Microsoft.DotNet.Interactive.AspNetCore.csproj", "{5167FC61-2FBF-4E0D-877E-C2DD354FB955}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "SampleLibrary", "src\SampleLibrary\SampleLibrary.csproj", "{E9EF70C0-AB48-4D7B-B978-517FE76BEFB0}"
```
Comment (jonsequitur) [inline dotnet-interactive.sln:79]: This library looks more like test support than a sample that would be of any interest to someone browsing the repo. I'd nest it under the test project.

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs [hunk 3]
```diff
@@ -116,6 +117,36 @@ public async Task language_service_methods_run_deferred_commands(Language langua
                 .ContainEquivalentOf(new FormattedValue(expectedMimeType, expectedContent));
         }
 
+        [Theory]
+        [InlineData(Language.CSharp, "var s = new SampleLibrary.Sample$$Library();")]
+        [InlineData(Language.FSharp, "let s = SampleLibrary.Sample$$Library()")]
+        public async Task hover_text_can_read_doc_comments_from_xml_files(Language language, string markupCode)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs:123]: Does this approach give us any assurances over:

* Docs for BCL types?
* Docs for types coming from NuGet packages?

Comment (brettfo) [reply src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs:123]: For (1), no; my initial PR comment tried to explain this.  The runtime assemblies don't have a .xml file next to them, so that'll be a separate issue to solve.  For (2), it depends on if the NuGet package _has_ the xml files.  If it does, then it'll get picked up, if not there's nothing we can do.

File: dotnet-interactive.sln (hunk 2 above)
Comment (brettfo) [reply dotnet-interactive.sln:79]: Not sure what you mean.  If you're asking about the shape in the Solution Explorer, it's already under test:
image

If you're asking about where it lives on disk, we don't have a separate `test/` directory, but I can certainly create one.

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs (hunk 3 above)
Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs:123]: Ah, sorry. I missed that comment and just dove right into the code in stereotypical engineer fashion. Sounds great.

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs (hunk 1 above)
Comment (brettfo) [reply src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs:122]: The `Skip` attribute isn't properly honored by VS test discovery, so this breaks my workflow.  Issue [here](https://github.com/xunit/xunit/issues/1782).

File: dotnet-interactive.sln (hunk 2 above)
Comment (jonsequitur) [reply dotnet-interactive.sln:79]: Mainly, the name is misleading. It's not a sample. Sometimes I put code like this inside the project folder, since it's really only of interest for a small subset of specific tests. 

Another approach that scales better is to use our test [utility](https://github.com/dotnet/interactive/blob/88a8ecb8fdd57513bea6eed3753be4da44568904/src/Microsoft.DotNet.Interactive.Tests/Utility/DirectoryUtility.cs#L48) that lets you declare file contents right in the test, and build this only when the test runs. It's more cohesive (you don't have to open up some other files to understand the test), more refactorable, and lets us add test cases more easily over time.

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs (hunk 3 above)
Comment (brettfo) [reply src/Microsoft.DotNet.Interactive.Tests/LanguageServices/HoverTextTests.cs:123]: I'm working on some tests for NuGet packages that I'll add to the PR.

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs (hunk 1 above)
Comment (brettfo) [reply src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs]: I've removed the F# test since it's known to not yet work via #1031.

File: dotnet-interactive.sln (hunk 2 above)
Comment (brettfo) [reply dotnet-interactive.sln]: Added a test helper that wraps that utility so sources can be specified on a per-test basis.

File: src/Microsoft.DotNet.Interactive.CSharp/CachingMetadataResolver.cs [hunk 4]
```diff
@@ -39,23 +40,28 @@ public override PortableExecutableReference ResolveMissingAssembly(MetadataRefer
...
             return new CachingMetadataResolver(SearchPaths, baseDirectory);
         }
 
         public ImmutableArray<string> SearchPaths => _resolver.SearchPaths;
 
         public string BaseDirectory => _resolver.BaseDirectory;
 
+        internal static PortableExecutableReference ResolveReferenceWithXmlDocumentationProvider(string path, MetadataReferenceProperties properties = default(MetadataReferenceProperties))
+        {
+            var peReference = MetadataReference.CreateFromFile(path, properties, XmlDocumentationProvider.CreateFromFile(Path.ChangeExtension(path, ".xml")));
```
Comment (brettfo) [inline src/Microsoft.DotNet.Interactive.CSharp/CachingMetadataResolver.cs:57]: Ultimately, this was the real fix; the rest was just plumbing.

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs [hunk 5]
```diff
@@ -115,5 +116,81 @@ public async Task signature_help_can_return_doc_comments_from_source(Language la
...
+                .Should()
+                .ContainSingle<SignatureHelpProduced>()
+                .Which
+                .Signatures
+                .Should()
+                .ContainSingle(sh => sh.Documentation.Value.Contains("Deserializes the JSON to a .NET object."));
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/interactive/issues/1071  N.b., the preceeding test can be deleted when this one is fixed.")]
+        public async Task csharp_signature_hep_can_read_doc_comments_from_nuget_packages()
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs:176]:
```suggestion
        public async Task csharp_signature_help_can_read_doc_comments_from_nuget_packages()
```

File: src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs [hunk 6]
```diff
@@ -115,5 +116,81 @@ public async Task signature_help_can_return_doc_comments_from_source(Language la
...
+                .Should()
+                .ContainSingle<SignatureHelpProduced>()
+                .Which
+                .Signatures
+                .Should()
+                .ContainSingle(sh => sh.Documentation.Value.Contains("A sample class constructor."));
+        }
+
+        [Fact]
+        public async Task csharp_signature_hep_can_read_doc_comments_from_nuget_packages_after_forcing_the_assembly_to_load()
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/LanguageServices/SignatureHelpTests.cs:152]:
```suggestion
        public async Task csharp_signature_help_can_read_doc_comments_from_nuget_packages_after_forcing_the_assembly_to_load()
```
